2021/day15/day15.java

```diff
@@ -0,0 +1,146 @@
+        try {
+            sc = new Scanner(new File(filename));
+            while (sc.hasNextLine()) {
+                ArrayList<Integer> line = new ArrayList<>();
+                sc.useDelimiter("");
+                while (sc.hasNextInt()) {
+                    line.add(sc.nextInt());
+                }
+                input.add(line);
+                if (sc.hasNextLine())
+                    sc.nextLine();
+            }
+        } catch (Exception e) {
+            System.out.println("Error handling input " + e);
+        } finally {
+            sc.close();
+        }
+        return input;
+    }
+
+    public static ArrayList<ArrayList<Integer>> modifyInput
+                (ArrayList<ArrayList<Integer>> mapOfRisks, int sizeMultiplier) {
+        for (ArrayList<Integer> existingLine : mapOfRisks) {
+            int startingLineSize = existingLine.size();
+            for (int i = 1; i < sizeMultiplier; i++) {
+                for (int y = 0; y < startingLineSize; y++) {
+                    int tmpRisk = existingLine.get(y) + i;
+                    if (tmpRisk > 9)
+                        tmpRisk = tmpRisk % 10 + 1;
+                    existingLine.add(tmpRisk);
+                }
+            }
+        }
+        int startingMapSize = mapOfRisks.size();
+        for (int i = 1; i < sizeMultiplier; i++) {
+            for (int y = 0; y < startingMapSize; y++) {
+                ArrayList<Integer> newLine = new ArrayList<>();
+                for (Integer risk : mapOfRisks.get(y)) {
+                    int tmpRisk = risk + i;
+                    if (tmpRisk > 9)
+                        tmpRisk = tmpRisk % 10 + 1;
+                    newLine.add(tmpRisk);
+                }
+                mapOfRisks.add(newLine);
+            }
+        }
+        return mapOfRisks;
+    }
+
+    /*
+     * Implementation of Dijkstra's algorithm for finding shortest path between
+     * two nodes. It goes like this:
+     * Do the following as long as we did not reach destination:
+     * 1. Visit all neighbour nodes (that are not in state "already visited")
+     * of the current node. Check their distance (risk) from the starting
+     * position and add them to priority queue. Set their position to
+     * "already visited".
+     * 2. From the priority queue take node that is the closest to starting
+     * position and move there.
+     * 3. Update current position and distance (risk).
+     */
+    public static int calculateRisk(boolean isPart2) {
+        ArrayList<ArrayList<Integer>> mapOfRisk = readInput("data.txt");
+        if (isPart2) {
+            mapOfRisk = modifyInput(mapOfRisk, 5);
+        }
+        ArrayList<ArrayList<Boolean>> isVisited = new ArrayList<>();
+        for (ArrayList<Integer> element : mapOfRisk) {
+            ArrayList<Boolean> line = new ArrayList<>
+                                (Collections.nCopies(mapOfRisk.size(), false));
+            isVisited.add(line);
+        }
```

Comment:
I like this, but just remember you could also use a `Set` and it could be a comparably fast and more space efficient